File: solution/2058. Find the Minimum and Maximum Number of Nodes Between Critical Points/2058.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    vector<int> nodesBetweenCriticalPoints(ListNode* head) {
        vector<int> res = {INT_MAX, -1};
        
        int min_c = 0, last = 0;
        int pre = 0;
        for (int i = 0; head && head->next; ++i, head = head->next) {
            int next = head->next->val;
            if ((pre < head->val && head->val > next) || (pre > head->val && head->val < next)) {
                if (last)
                    res[0] = min(res[0], i - last);
                if (!min_c)
                    min_c = i;
                last = i;
            }
            pre = head->val;
        }
        
        if (res[0] == INT_MAX) return {-1, -1};
        res[1] = last - min_c;
        return res;
    }
};

File: solution/2058. Find the Minimum and Maximum Number of Nodes Between Critical Points/2058.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def nodesBetweenCriticalPoints(self, head: Optional[ListNode]) -> List[int]:
        min_dist, max_dist = float('inf'), -1
        
        pre = 0
        first = last = i = 0
        while head and head.next:
            nxt = head.next.val
            
            if (pre > head.val < nxt) or (pre < head.val > nxt):
                if last:
                    min_dist = min(min_dist, i - last)
                if not first:
                    first = i
                last = i
            
            pre = head.val
            head = head.next
            i += 1
            
        if min_dist == float('inf'):
            return [-1, -1]
        max_dist = last - first
        return [min_dist, max_dist]
    
